File: app/src/main/java/com/mcdev/centercrop/CropActivity.java
package com.mcdev.centercrop;

import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;

import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.Toast;

import com.jackandphantom.blurimage.BlurImage;

import java.io.IOException;
import java.util.BitSet;

public class CropActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_crop);

        ImageView cropIV = findViewById(R.id.cropIV);
        ImageView blurIV = findViewById(R.id.blurIV);
        Button saveBtn = findViewById(R.id.saveBtn);
        ImageView discardIV = findViewById(R.id.discardIV);
        ConstraintLayout imageConstraintLayout = findViewById(R.id.imageConstraintLayout);

        if (getIntent() != null) {
            Intent intent = getIntent();
            String action = intent.getAction();
            Uri data = intent.getData();

            try {
                Bitmap bitmap = MediaStore.Images.Media.getBitmap(this.getContentResolver(), data);
                cropIV.setImageURI(data);
                BlurImage.with(this).load(bitmap).intensity(5).Async(true).into(blurIV);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }


        //save image
        saveBtn.setOnClickListener(v -> {
            imageConstraintLayout.setDrawingCacheEnabled(true);
            Bitmap b = imageConstraintLayout.getDrawingCache();
            MediaStore.Images.Media.insertImage(getApplicationContext().getContentResolver(), b, null, null);
            Toast.makeText(getApplicationContext(), "Image saved!", Toast.LENGTH_SHORT).show();
            this.finish();
        });

        //cancel
        discardIV.setOnClickListener(v -> {
            this.finish();
        });
    }
}

File: app/src/main/java/com/mcdev/jaspic/CropActivity.kt
package com.mcdev.jaspic

import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.provider.MediaStore
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.jackandphantom.blurimage.BlurImage
import com.mcdev.jaspic.databinding.ActivityCropBinding
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

class CropActivity : AppCompatActivity() {
    private val TAG = CropActivity::class.qualifiedName

    private lateinit var binding: ActivityCropBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCropBinding.inflate(layoutInflater)
        setContentView(binding.root)

        //getIntent
        if (intent != null) {
            val action = intent.action
            var data = intent.data

            if (action.equals(Intent.ACTION_SEND) && intent.type != null) {
                /*get Uri*/
                data = intent!!.getParcelableExtra(Intent.EXTRA_STREAM)
            }

            try {
                val bitmap = MediaStore.Images.Media.getBitmap(this.contentResolver, data)
                binding.cropIV.setImageURI(data)
                BlurImage.with(this).load(bitmap).intensity(5F).Async(true).into(binding.blurIV)
            } catch (e: IOException) {
                Log.d(TAG, "onCreate: $e")
            }

        }

        //save image
        binding.saveBtn.setOnClickListener {
            binding.imageConstraintLayout.isDrawingCacheEnabled = true
            val bitmap : Bitmap = binding.imageConstraintLayout.drawingCache
            saveImageToExternalStorage(bitmap)
            Toast.makeText(applicationContext, getString(R.string.image_saved), Toast.LENGTH_LONG).show()
            this.finish()
        }

        //discard changes
        binding.discardIV.setOnClickListener { this.finish() }
    }

    private fun refreshGallery(file : File) {
        MediaStore.Images.Media.insertImage(applicationContext.contentResolver, file.path, null, null) //did the magic. insert image in gallery
        intent = Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE)
        intent.data = Uri.fromFile(file)
        sendBroadcast(intent)
    }

    private fun saveImageToExternalStorage(finalBitmap: Bitmap) {
        val root = Environment.getExternalStorageDirectory()
        val myDir = File(root, "Jaspic")
        if (!myDir.exists()){
            myDir.mkdirs()
        }

        val filename = "Jaspic${System.currentTimeMillis()}.jpg"
        val file = File(myDir, filename)
        if (file.exists()) file.delete()
        try {
            val out = FileOutputStream(file)
            finalBitmap.compress(Bitmap.CompressFormat.JPEG, 100, out)
            refreshGallery(file)    //insert image to gallery and refresh
            out.flush()
            out.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}